File: src/commands/step.ts
import { readFileSync, writeFileSync } from "node:fs";
import * as yaml from "js-yaml";
import { HINTS } from "../lib/hints.js";
import { buildResponse, formatResponse } from "../lib/response-builder.js";
import { getSessionId, setSessionTask } from "../lib/session-manager.js";
import { findStep, resolveTaskFile } from "../lib/task-utils.js";
import { loadAndValidateTask, validateTask } from "../lib/validator.js";
import type { HistoryEvent, Step, TaskYaml } from "../types.js";

/**
 * Read stdin synchronously (for YAML input).
 */
function readStdin(): string {
	try {
		return readFileSync(0, "utf-8");
	} catch {
		return "";
	}
}

/**
 * Format steps list for output.
 */
function formatStepsList(steps: Step[]): Array<{
	id: string;
	goal: string;
	status: string;
	accept?: string | string[];
}> {
	return steps.map((s) => {
		const result: {
			id: string;
			goal: string;
			status: string;
			accept?: string | string[];
		} = {
			id: s.id,
			goal: s.goal,
			status: s.status ?? "pending",
		};
		if (s.accept) {
			result.accept = s.accept;
		}
		return result;
	});
}

/**
 * hal step add (single step mode)
 */
export function handleStepAdd(opts: {
	file?: string;
	step: string;
	goal: string;
	accept?: string[];
	after?: string;
}): void {
	const taskPath = resolveTaskFile(opts.file);
	const task = loadAndValidateTask(taskPath);

	if (!task.steps) {
		task.steps = [];
	}

	const existing = findStep(task.steps, opts.step);
	if (existing) {
		throw new Error(`ステップ ID "${opts.step}" は既に存在します`);
	}

	const newStep: Step = {
		id: opts.step,
		goal: opts.goal,
		status: "pending",
	};

	if (opts.accept && opts.accept.length > 0) {
		newStep.accept = opts.accept.length === 1 ? opts.accept[0] : opts.accept;
	}

	if (opts.after) {
		const afterIndex = task.steps.findIndex((s) => s.id === opts.after);
		if (afterIndex === -1) {
			throw new Error(`--after で指定されたステップ "${opts.after}" が見つかりません`);
		}
		task.steps.splice(afterIndex + 1, 0, newStep);
	} else {
		task.steps.push(newStep);
	}

	const now = new Date().toISOString();
	if (!task.history) {
		task.history = [];
	}
	const event: HistoryEvent = {
		at: now,
		type: "step_added",
		step: opts.step,
		message: `Step added: ${opts.goal}`,
	};
	task.history.push(event);

	validateTask(task);
	writeFileSync(taskPath, yaml.dump(task, { lineWidth: -1 }));

	const response = buildResponse({
		status: "ok",
		message: `ステップを追加しました: ${opts.step}`,
		data: {
			step_id: opts.step,
			goal: opts.goal,
			status: "pending",
		},
		commands_hint: HINTS.STEP_ADDED,
	});

	console.log(formatResponse(response));
}

/**
 * Input format for batch step add.
 */
interface StepInput {
	id: string;
	goal: string;
	accept?: string | string[];
}

/**
 * hal step add --stdin (batch mode)
 */
export function handleStepAddBatch(opts: { file?: string }): void {
	const input = readStdin().trim();
	if (!input) {
		throw new Error("stdin からステップデータを読み取れませんでした");
	}

	const stepsInput = yaml.load(input) as StepInput[];
	if (!Array.isArray(stepsInput) || stepsInput.length === 0) {
		throw new Error("stdin は YAML 配列形式で指定してください");
	}

	const taskPath = resolveTaskFile(opts.file);
	const task = loadAndValidateTask(taskPath);

	if (!task.steps) {
		task.steps = [];
	}

	const now = new Date().toISOString();
	if (!task.history) {
		task.history = [];
	}

	const inputIds = new Set<string>();
	for (const stepInput of stepsInput) {
		if (!stepInput.id || !stepInput.goal) {
			throw new Error(`ステップには id と goal が必要です: ${JSON.stringify(stepInput)}`);
		}
		if (inputIds.has(stepInput.id)) {
			throw new Error(`入力内でステップ ID "${stepInput.id}" が重複しています`);
		}
		inputIds.add(stepInput.id);
		const existing = findStep(task.steps, stepInput.id);
		if (existing) {
			throw new Error(`ステップ ID "${stepInput.id}" は既に存在します`);
		}
	}

	const addedSteps: string[] = [];

	for (const stepInput of stepsInput) {
		const newStep: Step = {
			id: stepInput.id,
			goal: stepInput.goal,
			status: "pending",
		};

		if (stepInput.accept) {
			newStep.accept = stepInput.accept;
		}

		task.steps.push(newStep);
		addedSteps.push(stepInput.id);

		const event: HistoryEvent = {
			at: now,
			type: "step_added",
			step: stepInput.id,
			message: `Step added: ${stepInput.goal}`,
		};
		task.history.push(event);
	}

	validateTask(task);
	writeFileSync(taskPath, yaml.dump(task, { lineWidth: -1 }));

	const response = buildResponse({
		status: "ok",
		message: `${addedSteps.length} ステップを追加しました`,
		data: {
			added: addedSteps,
			steps: formatStepsList(task.steps),
		},
		commands_hint: HINTS.STEP_ADDED,
	});

	console.log(formatResponse(response));
}

/**
 * hal step start
 *
 * Start working on a step. Updates session mapping.
 */
export function handleStepStart(opts: { file?: string; step: string }): void {
	const taskPath = resolveTaskFile(opts.file);
	const task = loadAndValidateTask(taskPath);

	if (!task.steps) {
		task.steps = [];
	}

	const step = findStep(task.steps, opts.step);
	if (!step) {
		throw new Error(`ステップ "${opts.step}" が見つかりません`);
	}

	const currentStatus = step.status ?? "pending";
	if (currentStatus !== "pending" && currentStatus !== "failed") {
		throw new Error(
			`ステップ "${opts.step}" は現在 ${currentStatus} です。pending または failed のステップのみ start できます`,
		);
	}

	step.status = "in_progress";

	if (task.status === "pending" || !task.status) {
		task.status = "in_progress";
	}

	const now = new Date().toISOString();
	if (!task.history) {
		task.history = [];
	}
	const event: HistoryEvent = {
		at: now,
		type: "step_started",
		step: opts.step,
	};
	task.history.push(event);

	validateTask(task);
	writeFileSync(taskPath, yaml.dump(task, { lineWidth: -1 }));

	// Update session mapping (supports cross-session handoff)
	try {
		const sessionId = getSessionId();
		setSessionTask(sessionId, taskPath);
	} catch {
		// No session ID — skip mapping
	}

	const responseData: Record<string, unknown> = {
		step_id: opts.step,
		step_goal: step.goal,
		step_status: "in_progress",
		task_goal: task.goal,
		task_status: task.status,
	};

	if (step.accept) {
		responseData.step_accept = step.accept;
	}

	const response = buildResponse({
		status: "ok",
		message: `ステップを開始しました: ${opts.step}`,
		data: responseData,
		commands_hint: HINTS.STEP_STARTED,
	});

	console.log(formatResponse(response));
}

/**
 * hal step done
 */
export function handleStepDone(opts: {
	file?: string;
	step: string;
	result: string;
	message: string;
}): void {
	const result = opts.result.toUpperCase();
	if (result !== "PASS" && result !== "FAIL") {
		throw new Error("--result は PASS または FAIL を指定してください");
	}

	const taskPath = resolveTaskFile(opts.file);
	const task = loadAndValidateTask(taskPath);

	if (!task.steps) {
		task.steps = [];
	}

	const step = findStep(task.steps, opts.step);
	if (!step) {
		throw new Error(`ステップ "${opts.step}" が見つかりません`);
	}

	const currentStatus = step.status ?? "pending";
	if (currentStatus !== "in_progress") {
		throw new Error(
			`ステップ "${opts.step}" は現在 ${currentStatus} です。in_progress のステップのみ done にできます`,
		);
	}

	if (result === "PASS" && step.accept && !step.verified) {
		throw new Error(
			`ステップ "${opts.step}" は未検証です。先にサブエージェントで hal step verify --step ${opts.step} --result PASS|FAIL を実行してください`,
		);
	}

	const now = new Date().toISOString();
	if (!task.history) {
		task.history = [];
	}

	if (result === "PASS") {
		step.status = "done";
		const event: HistoryEvent = {
			at: now,
			type: "step_done",
			step: opts.step,
			message: opts.message,
		};
		task.history.push(event);
	} else {
		const event: HistoryEvent = {
			at: now,
			type: "step_failed",
			step: opts.step,
			message: opts.message,
		};
		task.history.push(event);
	}

	const allDone =
		task.steps.length > 0 &&
		task.steps.every((s) => s.status === "done");

	if (allDone) {
		task.status = "done";
		const completedEvent: HistoryEvent = {
			at: now,
			type: "completed",
			message: "All steps completed",
		};
		task.history.push(completedEvent);
	}

	validateTask(task);
	writeFileSync(taskPath, yaml.dump(task, { lineWidth: -1 }));

	const responseData: Record<string, unknown> = {
		step_id: opts.step,
		result,
		step_status: step.status,
		task_status: task.status,
	};

	if (opts.message) {
		responseData.message = opts.message;
	}

	let commandsHint: string;
	if (allDone) {
		commandsHint = HINTS.STEP_DONE_ALL;
	} else if (result === "FAIL") {
		commandsHint = HINTS.STEP_DONE_FAIL;
	} else {
		const nextStep = task.steps.find((s) => (s.status ?? "pending") === "pending");
		if (nextStep) {
			commandsHint = HINTS.STEP_DONE_NEXT(nextStep.id);
		} else {
			commandsHint = HINTS.STEP_DONE_CHECK_STATUS;
		}
	}

	const response = buildResponse({
		status: "ok",
		message:
			result === "PASS"
				? `ステップ完了: ${opts.step}`
				: `ステップ失敗: ${opts.step}`,
		data: responseData,
		commands_hint: commandsHint,
	});

	console.log(formatResponse(response));
}

/**
 * hal step pause
 */
export function handleStepPause(opts: { file?: string; message: string }): void {
	const taskPath = resolveTaskFile(opts.file);
	const task = loadAndValidateTask(taskPath);

	const now = new Date().toISOString();
	if (!task.history) {
		task.history = [];
	}
	const event: HistoryEvent = {
		at: now,
		type: "paused",
		message: opts.message,
	};
	task.history.push(event);

	validateTask(task);
	writeFileSync(taskPath, yaml.dump(task, { lineWidth: -1 }));

	const currentStep = task.steps?.find((s) => s.status === "in_progress");

	const responseData: Record<string, unknown> = {
		task_status: task.status,
		paused: true,
	};

	if (currentStep) {
		responseData.current_step = currentStep.id;
		responseData.step_goal = currentStep.goal;
	}

	responseData.pause_reason = opts.message;

	const response = buildResponse({
		status: "ok",
		message: `作業を一時停止しました: ${opts.message}`,
		data: responseData,
		commands_hint: HINTS.STEP_PAUSED,
	});

	console.log(formatResponse(response));
}

/**
 * hal step resume
 */
export function handleStepResume(opts: { file?: string }): void {
	const taskPath = resolveTaskFile(opts.file);
	const task = loadAndValidateTask(taskPath);

	const now = new Date().toISOString();
	if (!task.history) {
		task.history = [];
	}
	const event: HistoryEvent = {
		at: now,
		type: "resumed",
		message: "Work resumed",
	};
	task.history.push(event);

	validateTask(task);
	writeFileSync(taskPath, yaml.dump(task, { lineWidth: -1 }));

	const inProgressSteps = (task.steps ?? []).filter((s) => s.status === "in_progress");
	const pendingSteps = (task.steps ?? []).filter((s) => s.status === "pending" || !s.status);

	const data: Record<string, unknown> = {
		task_status: task.status ?? "pending",
	};

	if (inProgressSteps.length > 0) {
		data.current_step = {
			id: inProgressSteps[0].id,
			goal: inProgressSteps[0].goal,
		};
	} else if (pendingSteps.length > 0) {
		data.next_step = {
			id: pendingSteps[0].id,
			goal: pendingSteps[0].goal,
		};
	}

	data.steps = formatStepsList(task.steps ?? []);

	const response = buildResponse({
		status: "ok",
		message: "作業を再開しました",
		data,
		commands_hint: HINTS.STEP_RESUMED,
	});

	console.log(formatResponse(response));
}

/**
 * hal step verify
 */
export function handleStepVerify(opts: {
	file?: string;
	step: string;
	result: string;
	message: string;
}): void {
	const taskPath = resolveTaskFile(opts.file);
	const task = loadAndValidateTask(taskPath);

	if (!task.steps || task.steps.length === 0) {
		throw new Error("タスクにステップがありません");
	}

	const step = findStep(task.steps, opts.step);
	if (!step) {
		throw new Error(`ステップ "${opts.step}" が見つかりません`);
	}

	const result = opts.result.toUpperCase();
	if (result !== "PASS" && result !== "FAIL") {
		throw new Error('result は PASS または FAIL を指定してください');
	}

	step.verified = result === "PASS";

	const now = new Date().toISOString();
	if (!task.history) {
		task.history = [];
	}
	const event: HistoryEvent = {
		at: now,
		type: "step_verified",
		step: opts.step,
		result: result as "PASS" | "FAIL",
		message: opts.message,
	};
	task.history.push(event);

	validateTask(task);
	writeFileSync(taskPath, yaml.dump(task, { lineWidth: -1 }));

	const response = buildResponse({
		status: "ok",
		message: result === "PASS"
			? `検証完了: ステップ ${opts.step} は PASS`
			: `検証失敗: ステップ ${opts.step} は FAIL`,
		data: {
			step_id: opts.step,
			result,
			verified: step.verified,
		},
		commands_hint: result === "PASS"
			? HINTS.STEP_DONE_NEXT(opts.step)
			: HINTS.STEP_DONE_FAIL,
	});

	console.log(formatResponse(response));
}

File: src/commands/step.rs
use anyhow::{Result, bail};
use std::collections::BTreeMap;
use std::io::Read;

use crate::core::hints;
use crate::core::response::HalResponse;
use crate::core::session::{get_session_id, set_session_task};
use crate::core::task_utils::resolve_task_file;
use crate::core::validator::{load_and_validate_task, save_task};
use crate::types::{AcceptCriteria, HistoryEvent, Status, Step};

fn find_step_index(steps: &[Step], step_id: &str) -> Option<usize> {
    steps.iter().position(|s| s.id == step_id)
}

fn now() -> String {
    chrono::Utc::now().to_rfc3339()
}

fn yaml_str(s: &str) -> serde_yaml::Value {
    serde_yaml::Value::String(s.to_string())
}

pub fn handle_step_add(
    file: Option<&str>, step_id: &str, goal: &str, accept: &[String], after: Option<&str>,
) -> Result<()> {
    let task_path = resolve_task_file(file)?;
    let mut task = load_and_validate_task(&task_path)?;
    let steps = task.steps.get_or_insert_with(Vec::new);

    if find_step_index(steps, step_id).is_some() {
        bail!("Step ID \"{}\" already exists", step_id);
    }

    let mut new_step = Step {
        id: step_id.to_string(),
        goal: goal.to_string(),
        status: Some(Status::Pending),
        accept: None,
        verified: None,
    };

    if !accept.is_empty() {
        new_step.accept = Some(if accept.len() == 1 {
            AcceptCriteria::Single(accept[0].clone())
        } else {
            AcceptCriteria::Multiple(accept.to_vec())
        });
    }

    if let Some(after_id) = after {
        let idx = find_step_index(steps, after_id)
            .ok_or_else(|| anyhow::anyhow!("Step \"{}\" specified in --after not found", after_id))?;
        steps.insert(idx + 1, new_step);
    } else {
        steps.push(new_step);
    }

    let history = task.history.get_or_insert_with(Vec::new);
    history.push(HistoryEvent::StepAdded {
        at: now(),
        step: step_id.to_string(),
        message: Some(format!("Step added: {}", goal)),
    });

    save_task(&task_path, &task)?;

    let mut data = BTreeMap::new();
    data.insert("step_id".into(), yaml_str(step_id));
    data.insert("goal".into(), yaml_str(goal));
    data.insert("status".into(), yaml_str("pending"));

    let response = HalResponse::new("ok", &format!("Step added: {}", step_id))
        .with_data(data)
        .with_hint(hints::STEP_ADDED);
    println!("{}", response.to_yaml());
    Ok(())
}

pub fn handle_step_add_batch(file: Option<&str>) -> Result<()> {
    let mut input = String::new();
    std::io::stdin().read_to_string(&mut input)?;
    let input = input.trim();
    if input.is_empty() {
        bail!("Failed to read step data from stdin");
    }

    let steps_input: Vec<serde_yaml::Value> = serde_yaml::from_str(input)?;
    if steps_input.is_empty() {
        bail!("stdin must be a YAML array");
    }

    let task_path = resolve_task_file(file)?;
    let mut task = load_and_validate_task(&task_path)?;
    let steps = task.steps.get_or_insert_with(Vec::new);
    let history = task.history.get_or_insert_with(Vec::new);
    let ts = now();

    let mut added = vec![];
    let mut seen_ids = std::collections::HashSet::new();

    for entry in &steps_input {
        let id = entry.get("id").and_then(|v| v.as_str())
            .ok_or_else(|| anyhow::anyhow!("Step requires id and goal"))?;
        let goal = entry.get("goal").and_then(|v| v.as_str())
            .ok_or_else(|| anyhow::anyhow!("Step requires id and goal"))?;

        if !seen_ids.insert(id.to_string()) {
            bail!("Duplicate step ID in input: \"{}\"", id);
        }
        if find_step_index(steps, id).is_some() {
            bail!("Step ID \"{}\" already exists", id);
        }

        let accept = entry.get("accept").map(|v| match v {
            serde_yaml::Value::String(s) => AcceptCriteria::Single(s.clone()),
            serde_yaml::Value::Sequence(arr) => {
                AcceptCriteria::Multiple(arr.iter().filter_map(|x| x.as_str().map(String::from)).collect())
            }
            _ => AcceptCriteria::Single(format!("{:?}", v)),
        });

        steps.push(Step {
            id: id.to_string(),
            goal: goal.to_string(),
            status: Some(Status::Pending),
            accept,
            verified: None,
        });

        history.push(HistoryEvent::StepAdded {
            at: ts.clone(),
            step: id.to_string(),
            message: Some(format!("Step added: {}", goal)),
        });

        added.push(id.to_string());
    }

    save_task(&task_path, &task)?;

    let mut data = BTreeMap::new();
    let added_yaml: Vec<serde_yaml::Value> = added.iter().map(|s| yaml_str(s)).collect();
    data.insert("added".into(), serde_yaml::Value::Sequence(added_yaml));

    let response = HalResponse::new("ok", &format!("{} steps added", added.len()))
        .with_data(data)
        .with_hint(hints::STEP_ADDED);
    println!("{}", response.to_yaml());
    Ok(())
}

pub fn handle_step_start(file: Option<&str>, step_id: &str) -> Result<()> {
    let task_path = resolve_task_file(file)?;
    let mut task = load_and_validate_task(&task_path)?;

    // Validate and mutate steps
    let step_goal = {
        let steps = task.steps.get_or_insert_with(Vec::new);
        let idx = find_step_index(steps, step_id)
            .ok_or_else(|| anyhow::anyhow!("Step \"{}\" not found", step_id))?;
        let current = steps[idx].status.as_ref().unwrap_or(&Status::Pending);
        if *current != Status::Pending && *current != Status::Failed {
            bail!("Step \"{}\" is currently {}. Only pending or failed steps can be started", step_id, current);
        }
        steps[idx].status = Some(Status::InProgress);
        steps[idx].goal.clone()
    };

    if task.status.as_ref().unwrap_or(&Status::Pending) == &Status::Pending {
        task.status = Some(Status::InProgress);
    }

    let history = task.history.get_or_insert_with(Vec::new);
    history.push(HistoryEvent::StepStarted {
        at: now(),
        step: step_id.to_string(),
        message: None,
    });

    save_task(&task_path, &task)?;

    if let Ok(session_id) = get_session_id() {
        let _ = set_session_task(&session_id, task_path.to_str().unwrap_or(""));
    }

    let mut data = BTreeMap::new();
    data.insert("step_id".into(), yaml_str(step_id));
    data.insert("step_goal".into(), yaml_str(&step_goal));
    data.insert("step_status".into(), yaml_str("in_progress"));
    data.insert("task_goal".into(), yaml_str(&task.goal));
    data.insert("task_status".into(), yaml_str(&task.status.as_ref().unwrap().to_string()));

    let response = HalResponse::new("ok", &format!("Step started: {}", step_id))
        .with_data(data)
        .with_hint(hints::STEP_STARTED);
    println!("{}", response.to_yaml());
    Ok(())
}

pub fn handle_step_done(file: Option<&str>, step_id: &str, result: &str, message: &str) -> Result<()> {
    let result_upper = result.to_uppercase();
    if result_upper != "PASS" && result_upper != "FAIL" {
        bail!("--result must be PASS or FAIL");
    }

    let task_path = resolve_task_file(file)?;
    let mut task = load_and_validate_task(&task_path)?;

    // Validate and mutate steps in a block to release borrow
    {
        let steps = task.steps.get_or_insert_with(Vec::new);
        let idx = find_step_index(steps, step_id)
            .ok_or_else(|| anyhow::anyhow!("Step \"{}\" not found", step_id))?;

        let current = steps[idx].status.as_ref().unwrap_or(&Status::Pending);
        if *current != Status::InProgress {
            bail!("Step \"{}\" is currently {}. Only in_progress steps can be marked done", step_id, current);
        }

        if result_upper == "PASS" && steps[idx].accept.is_some() && steps[idx].verified != Some(true) {
            bail!("Step \"{}\" is unverified. Run hal step verify --step {} --result PASS|FAIL via sub-agent first", step_id, step_id);
        }

        if result_upper == "PASS" {
            steps[idx].status = Some(Status::Done);
        }
    }

    let ts = now();
    let history = task.history.get_or_insert_with(Vec::new);

    if result_upper == "PASS" {
        history.push(HistoryEvent::StepDone {
            at: ts.clone(),
            step: step_id.to_string(),
            message: Some(message.to_string()),
        });
    } else {
        history.push(HistoryEvent::StepFailed {
            at: ts.clone(),
            step: step_id.to_string(),
            message: Some(message.to_string()),
        });
    }

    let steps = task.steps.as_ref().unwrap();
    let all_done = !steps.is_empty() && steps.iter().all(|s| s.status.as_ref() == Some(&Status::Done));
    if all_done {
        task.status = Some(Status::Done);
        let history = task.history.get_or_insert_with(Vec::new);
        history.push(HistoryEvent::Completed {
            at: ts,
            message: Some("All steps completed".to_string()),
        });
    }

    save_task(&task_path, &task)?;

    let steps = task.steps.as_ref().unwrap();
    let step_status = steps.iter().find(|s| s.id == step_id).unwrap().status.as_ref().unwrap().to_string();

    let mut data = BTreeMap::new();
    data.insert("step_id".into(), yaml_str(step_id));
    data.insert("result".into(), yaml_str(&result_upper));
    data.insert("step_status".into(), yaml_str(&step_status));
    data.insert("task_status".into(), yaml_str(&task.status.as_ref().unwrap_or(&Status::Pending).to_string()));

    let hint = if all_done {
        hints::STEP_DONE_ALL.to_string()
    } else if result_upper == "FAIL" {
        hints::STEP_DONE_FAIL.to_string()
    } else {
        steps.iter()
            .find(|s| s.status.as_ref().unwrap_or(&Status::Pending) == &Status::Pending)
            .map(|s| hints::step_done_next(&s.id))
            .unwrap_or_else(|| hints::STEP_DONE_CHECK_STATUS.to_string())
    };

    let msg = if result_upper == "PASS" {
        format!("Step completed: {}", step_id)
    } else {
        format!("Step failed: {}", step_id)
    };

    let response = HalResponse::new("ok", &msg).with_data(data).with_hint(&hint);
    println!("{}", response.to_yaml());
    Ok(())
}

pub fn handle_step_pause(file: Option<&str>, message: &str) -> Result<()> {
    let task_path = resolve_task_file(file)?;
    let mut task = load_and_validate_task(&task_path)?;

    let history = task.history.get_or_insert_with(Vec::new);
    history.push(HistoryEvent::Paused {
        at: now(),
        message: Some(message.to_string()),
    });

    save_task(&task_path, &task)?;

    let mut data = BTreeMap::new();
    data.insert("task_status".into(), yaml_str(&task.status.as_ref().unwrap_or(&Status::Pending).to_string()));
    data.insert("paused".into(), serde_yaml::Value::Bool(true));
    data.insert("pause_reason".into(), yaml_str(message));

    let response = HalResponse::new("ok", &format!("Work paused: {}", message))
        .with_data(data)
        .with_hint(hints::STEP_PAUSED);
    println!("{}", response.to_yaml());
    Ok(())
}

pub fn handle_step_resume(file: Option<&str>) -> Result<()> {
    let task_path = resolve_task_file(file)?;
    let mut task = load_and_validate_task(&task_path)?;

    let history = task.history.get_or_insert_with(Vec::new);
    history.push(HistoryEvent::Resumed {
        at: now(),
        message: Some("Work resumed".to_string()),
    });

    save_task(&task_path, &task)?;

    let mut data = BTreeMap::new();
    data.insert("task_status".into(), yaml_str(&task.status.as_ref().unwrap_or(&Status::Pending).to_string()));

    let response = HalResponse::new("ok", "Work resumed")
        .with_data(data)
        .with_hint(hints::STEP_RESUMED);
    println!("{}", response.to_yaml());
    Ok(())
}

pub fn handle_step_verify(file: Option<&str>, step_id: &str, result: &str, message: &str) -> Result<()> {
    let result_upper = result.to_uppercase();
    if result_upper != "PASS" && result_upper != "FAIL" {
        bail!("--result must be PASS or FAIL");
    }

    let task_path = resolve_task_file(file)?;
    let mut task = load_and_validate_task(&task_path)?;

    let verified = {
        let steps = task.steps.as_mut().ok_or_else(|| anyhow::anyhow!("Task has no steps"))?;
        let idx = find_step_index(steps, step_id)
            .ok_or_else(|| anyhow::anyhow!("Step \"{}\" not found", step_id))?;
        steps[idx].verified = Some(result_upper == "PASS");
        steps[idx].verified.unwrap_or(false)
    };

    let history = task.history.get_or_insert_with(Vec::new);
    history.push(HistoryEvent::StepVerified {
        at: now(),
        step: step_id.to_string(),
        result: result_upper.clone(),
        message: Some(message.to_string()),
    });

    save_task(&task_path, &task)?;

    let mut data = BTreeMap::new();
    data.insert("step_id".into(), yaml_str(step_id));
    data.insert("result".into(), yaml_str(&result_upper));
    data.insert("verified".into(), serde_yaml::Value::Bool(verified));

    let hint = if result_upper == "PASS" {
        hints::step_done_next(step_id)
    } else {
        hints::STEP_DONE_FAIL.to_string()
    };

    let msg = if result_upper == "PASS" {
        format!("Verification passed: step {}", step_id)
    } else {
        format!("Verification failed: step {}", step_id)
    };

    let response = HalResponse::new("ok", &msg).with_data(data).with_hint(&hint);
    println!("{}", response.to_yaml());
    Ok(())
}
